File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        unordered_map<char, string> combinations = {
            {'2', "abc"}, {'3', "def"},  {'4', "ghi"}, {'5', "jkl"},
            {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}};

        string current;
        vector<string> answer;
        const int n = digits.size();

        if (n == 0)
            return answer;
        backtrack(0, digits, combinations, current, answer, n);
        return answer;
    }

    void backtrack(int index, string& digits,
                   unordered_map<char, string>& combinations, string& current,
                   vector<string>& answer, int n) {
        if (index == n) {
            answer.push_back(current);
            return;
        }

        char digit = digits[index];
        string letters = combinations.at(digit);

        for (auto c : letters) {
            current.push_back(c);
            backtrack(index + 1, digits, combinations, current, answer, n);
            current.pop_back();
        }
    }
};

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        phone = {
            "2": ["a", "b", "c"],
            "3": ["d", "e", "f"],
            "4": ["g", "h", "i"],
            "5": ["j", "k", "l"],
            "6": ["m", "n", "o"],
            "7": ["p", "q", "r", "s"],
            "8": ["t", "u", "v"],
            "9": ["w", "x", "y", "z"]
        }

        ans = []

        if not digits:
            return ans
        
        # backtrack
        def backtrack(indx,temp_ans ):
            # base condition
            if indx >= len(digits):
                ans.append("".join(temp_ans))
                return

            # main logic
            digit = digits[indx] # 2
            letters = phone[digit]

            # loop
            for i in letters:
                temp_ans.append(i)
                backtrack(indx+1, temp_ans)
                temp_ans.pop()
            
        backtrack(0, [])
        
        return ans





